src/box/lua/config/configdata.lua

```diff
@@ -102,6 +102,19 @@ local function instance_sharding(iconfig, instance_name)
         u.login = 'guest'
         uri = urilib.format(u, true)
     end
+
+    -- Check that the vshard storage user has the credential sharding role.
+    local has_sharding = false
+    local user_roles = instance_config:get(iconfig, {'credentials', 'users',
+                                                     u.login, 'roles'}) or {}
+    for _, cred_role in pairs(user_roles) do
+        has_sharding = has_sharding or cred_role == 'sharding'
+    end
+    if not has_sharding then
+        local err = "storage user %q should have %q role"
+        error(err:format(u.login, 'sharding'), 0)
+    end
```

Comment:
Shouldn't we allow to set a user in `iproto.advertise.sharding` that is not present in the `credentials` section at all? I mean, it may be present in snapshots already and have all the needed privileges.

Reply:
True. Fixed.